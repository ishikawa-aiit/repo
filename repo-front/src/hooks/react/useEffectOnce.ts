import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
