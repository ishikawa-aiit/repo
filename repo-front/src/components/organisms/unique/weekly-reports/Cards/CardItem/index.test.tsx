import { render, screen, within } from '@testing-library/react'

import { CardItem } from '.'

const testData = {
  title: 'テストタイトル',
  description: 'テスト詳細文',
  isCopyable: false,
}

describe('CardItem', () => {
  it('正常にレンダリングされている', () => {
    const { title, description } = testData
    const { container } = render(<CardItem {...testData} />)
    const { getByRole } = within(screen.getByTestId('CardItem'))

    expect(getByRole('term')).toHaveTextContent(title)
    expect(getByRole('definition')).toHaveTextContent(description)
    expect(container).toMatchSnapshot()
  })
})
