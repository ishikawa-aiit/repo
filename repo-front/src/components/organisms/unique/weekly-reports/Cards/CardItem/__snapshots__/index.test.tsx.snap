// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItem 正常にレンダリングされている 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-bottom: 1px;
  border-color: neutral.200;
  width: 100%;
  word-break: break-all;
  padding-top: 2px;
  padding-bottom: 4px;
  gap: 4px;
}

.emotion-0:last-of-type {
  border-bottom: none;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  font-weight: bold;
}

.emotion-2 {
  vertical-align: middle;
  white-space: pre-wrap;
}

<div>
  <dl
    class="emotion-0"
    data-cy="CardItem"
    data-testid="CardItem"
  >
    <dt
      aria-label="title"
      class="emotion-1"
    >
      テストタイトル
    </dt>
    <dd
      class="chakra-text emotion-2"
    >
      テスト詳細文
    </dd>
  </dl>
</div>
`;
